src/net/tls.cc

```diff
@@ -1217,7 +1223,11 @@ class session : public enable_lw_shared_from_this<session> {
             gnutls_x509_crt_deinit(peer_leaf_cert);
 
             if (ec || ec2) {
-                throw std::runtime_error("error while extracting certificate DN strings");
+                std::runtime_error err("error while extracting certificate DN strings");
+                if (fetch_dn) {
+                    _session_dn->set_exception(err);
```

Comment:
what happens if `fetch_dn` is set, but other paths return before the promise is set. for example, above `if (fetch_dn || _creds->_dn_callback) {` I see three extra return paths (1 return and 2 throws).

Reply:
fixed that